File: html5/src/chibi/ChibiClips.js
// -----------------------------------------------------------------------------
// ChibiClips — the motion, in two halves.
//
// Locomotion is procedural: idle, walk and run are functions of phase, so a walk
// cycle can scale with how fast the actor is actually moving and a game needs no
// files to have a character that is alive. Everything expressive is keyed, out of
// chibi-clips.json, because a wave is a performance and a sine wave is not.
//
// Both halves produce the same thing — a Pose — so the animator blends between a
// procedural clip and a keyed one without caring which is which.
//
// Mirrored in SexyBiscuit.Engine/Chibi/ChibiClips.cs.
// -----------------------------------------------------------------------------

import keyed from './chibi-clips.json' with { type: 'json' };

const TAU = Math.PI * 2;

/**
 * A pose: a rotation per joint in euler degrees, plus a displacement of the whole
 * character. Joints left out of a pose stay at rest, which is what lets a wave
 * animate one arm and leave the legs alone.
 */
export class Pose {
    constructor() {
        /** @type {Map<string, [number, number, number]>} */
        this.joints = new Map();
        this.offset = [0, 0, 0];
    }

    set(joint, x, y, z) { this.joints.set(joint, [x, y, z]); return this; }

    clear() { this.joints.clear(); this.offset = [0, 0, 0]; return this; }

    /** Reads a joint's rotation, or the rest pose when the clip did not set one. */
    get(joint) { return this.joints.get(joint) ?? ZERO; }
}

const ZERO = Object.freeze([0, 0, 0]);

/** Every joint either side of a blend, so neither clip's joints snap. */
export function blendInto(out, a, b, amount) {
    out.clear();
    const names = new Set([...a.joints.keys(), ...b.joints.keys()]);
    for (const name of names) {
        const from = a.get(name);
        const to = b.get(name);
        out.set(name,
            from[0] + (to[0] - from[0]) * amount,
            from[1] + (to[1] - from[1]) * amount,
            from[2] + (to[2] - from[2]) * amount);
    }
    for (let axis = 0; axis < 3; axis++) {
        out.offset[axis] = a.offset[axis] + (b.offset[axis] - a.offset[axis]) * amount;
    }
    return out;
}

// -----------------------------------------------------------------------------
// Procedural locomotion
// -----------------------------------------------------------------------------

/** Arms hang a little away from the body; dead-straight arms read as a mannequin. */
const ARM_REST = 7;
/** And the elbows keep a slight bend for the same reason. */
const ELBOW_REST = -12;

function idle(t, pose, intensity) {
    const b = Math.sin(t * TAU) * intensity;

    pose.set('Torso', 1.5 + b * 1.6, 0, 0);
    pose.set('Head', -b * 2, Math.sin(t * TAU * 0.5) * 4, 0);
    pose.set('ArmL', b * 1.5, 0, ARM_REST + b);
    pose.set('ArmR', b * 1.5, 0, -ARM_REST - b);
    pose.set('ForearmL', ELBOW_REST, 0, 0);
    pose.set('ForearmR', ELBOW_REST, 0, 0);
    pose.offset[1] = b * 0.006;
}

function stride(t, pose, swing, knee, lean, bob, elbow) {
    const s = Math.sin(t * TAU);
    const c = Math.cos(t * TAU);

    const thighL = s * swing;
    const thighR = -s * swing;
    // The knee only folds on the back half of the stride; a leg that bends going
    // forward is the single thing that makes a walk cycle look wrong.
    const shinL = -Math.max(0, -s) * knee;
    const shinR = -Math.max(0, s) * knee;

    pose.set('ThighL', thighL, 0, 0);
    pose.set('ThighR', thighR, 0, 0);
    pose.set('ShinL', shinL, 0, 0);
    pose.set('ShinR', shinR, 0, 0);
    // Feet stay roughly flat by undoing half of what the leg above them did.
    pose.set('FootL', -(thighL + shinL) * 0.5, 0, 0);
    pose.set('FootR', -(thighR + shinR) * 0.5, 0, 0);

    pose.set('ArmL', -s * swing * 0.85, 0, ARM_REST);
    pose.set('ArmR', s * swing * 0.85, 0, -ARM_REST);
    pose.set('ForearmL', elbow, 0, 0);
    pose.set('ForearmR', elbow, 0, 0);

    pose.set('Hips', 0, -s * 6, 0);
    pose.set('Torso', lean, s * 5, 0);
    pose.set('Head', -lean * 0.5, -s * 3, 0);

    // Two bobs per stride: one for each foot that lands.
    pose.offset[1] = Math.abs(c) * bob;
}

/**
 * The procedural clips. `intensity` is 1 at the natural speed and scales the
 * whole motion, so a character shuffling at half speed swings its arms half as
 * far instead of playing the same cycle slowly.
 */
export const PROCEDURAL = Object.freeze({
    idle: { duration: 3.2, loop: true, sample: (t, pose, k = 1) => idle(t, pose, k) },
    walk: {
        duration: 0.86, loop: true,
        sample: (t, pose, k = 1) => stride(t, pose, 30 * k, 34 * k, 3 * k, 0.018 * k, ELBOW_REST),
    },
    run: {
        duration: 0.56, loop: true,
        sample: (t, pose, k = 1) => stride(t, pose, 48 * k, 62 * k, 14 * k, 0.032 * k, -58),
    },
});

// -----------------------------------------------------------------------------
// Keyed clips
// -----------------------------------------------------------------------------

/** The raw keyed clip data, as loaded. */
export const KEYED = keyed.clips;

function sampleTrack(keys, t) {
    if (keys.length === 0) return ZERO;
    if (t <= keys[0].t) return keys[0].rot ?? keys[0].pos;

    for (let i = 1; i < keys.length; i++) {
        if (t > keys[i].t) continue;
        const a = keys[i - 1];
        const b = keys[i];
        const span = b.t - a.t;
        const k = span <= 0 ? 0 : (t - a.t) / span;
        const from = a.rot ?? a.pos;
        const to = b.rot ?? b.pos;
        return [
            from[0] + (to[0] - from[0]) * k,
            from[1] + (to[1] - from[1]) * k,
            from[2] + (to[2] - from[2]) * k,
        ];
    }
    const last = keys[keys.length - 1];
    return last.rot ?? last.pos;
}

/** Writes a keyed clip's pose at normalised time `t` into `pose`. */
export function sampleKeyed(clip, t, pose) {
    for (const [joint, keys] of Object.entries(clip.tracks)) {
        const rot = sampleTrack(keys, t);
        pose.set(joint, rot[0], rot[1], rot[2]);
    }
    if (clip.offset) pose.offset = sampleTrack(clip.offset, t).slice();
}

/** Every clip name a game may play, procedural and keyed together. */
export function clipNames() {
    return [...Object.keys(PROCEDURAL), ...Object.keys(KEYED)];
}

/**
 * Looks a clip up by name, in a uniform shape.
 * @returns {?{name, duration, loop, hold, sample(t, pose, intensity)}}
 */
export function findClip(name) {
    const procedural = PROCEDURAL[name];
    if (procedural) {
        return { name, duration: procedural.duration, loop: procedural.loop, hold: false,
                 sample: procedural.sample };
    }

    const clip = KEYED[name];
    if (!clip) return null;
    return {
        name,
        duration: clip.duration,
        loop: clip.loop === true,
        hold: clip.hold === true,
        sample: (t, pose) => sampleKeyed(clip, t, pose),
    };
}

File: SexyBiscuit.Engine/Chibi/ChibiClips.cs
using System.Text.Json;
using Microsoft.Xna.Framework;

namespace SexyBiscuit.Engine.Chibi;

/// <summary>
/// A pose: a rotation per joint in euler degrees, plus a displacement of the whole
/// character.
/// </summary>
/// <remarks>
/// Joints left out of a pose stay at rest, which is what lets a wave animate one arm and
/// leave the legs alone.
/// </remarks>
public sealed class ChibiPose
{
    /// <summary>Joint name to its euler rotation in degrees.</summary>
    public Dictionary<string, Vector3> Joints { get; } = new(StringComparer.Ordinal);

    /// <summary>How far the whole character is displaced, for a bob or a jump.</summary>
    public Vector3 Offset { get; set; }

    /// <summary>Sets one joint's rotation.</summary>
    public ChibiPose Set(string joint, float x, float y, float z)
    {
        Joints[joint] = new Vector3(x, y, z);
        return this;
    }

    /// <summary>Empties the pose, returning every joint to rest.</summary>
    public ChibiPose Clear()
    {
        Joints.Clear();
        Offset = Vector3.Zero;
        return this;
    }

    /// <summary>One joint's rotation, or zero when the clip did not set it.</summary>
    public Vector3 Get(string joint)
        => Joints.TryGetValue(joint, out Vector3 rotation) ? rotation : Vector3.Zero;

    /// <summary>Blends two poses into a third, covering every joint either of them names.</summary>
    public static ChibiPose Blend(ChibiPose into, ChibiPose from, ChibiPose to, float amount)
    {
        into.Clear();
        foreach (string name in from.Joints.Keys.Concat(to.Joints.Keys).Distinct())
            into.Joints[name] = Vector3.Lerp(from.Get(name), to.Get(name), amount);

        into.Offset = Vector3.Lerp(from.Offset, to.Offset, amount);
        return into;
    }
}

/// <summary>One clip, whether it came from code or from the shared file.</summary>
/// <param name="Name">The name a game plays it by.</param>
/// <param name="Duration">Seconds at speed 1.</param>
/// <param name="Loop">Whether it repeats rather than ending.</param>
/// <param name="Hold">Whether it keeps its last pose instead of returning to rest.</param>
/// <param name="Sample">Writes the pose at normalised time into the given pose.</param>
public sealed record ChibiClip(
    string Name, float Duration, bool Loop, bool Hold,
    Action<float, ChibiPose, float> Sample);

/// <summary>
/// The motion, in two halves.
/// </summary>
/// <remarks>
/// <para>
/// Locomotion is procedural: idle, walk and run are functions of phase, so a walk cycle can
/// scale with how fast the actor is actually moving and a game needs no files to have a
/// character that is alive. Everything expressive is keyed, out of
/// <c>html5/src/chibi/chibi-clips.json</c>, because a wave is a performance and a sine wave
/// is not.
/// </para>
/// <para>Mirrored in <c>html5/src/chibi/ChibiClips.js</c>.</para>
/// </remarks>
public static class ChibiClips
{
    private const float Tau = MathF.PI * 2f;

    /// <summary>Arms hang a little away from the body; dead-straight arms read as a mannequin.</summary>
    private const float ArmRest = 7f;

    /// <summary>And the elbows keep a slight bend for the same reason.</summary>
    private const float ElbowRest = -12f;

    private static readonly Dictionary<string, ChibiClip> _keyed = LoadKeyed();

    private static readonly Dictionary<string, ChibiClip> _procedural = new(StringComparer.Ordinal)
    {
        ["idle"] = new("idle", 3.2f, true, false, Idle),
        ["walk"] = new("walk", 0.86f, true, false,
            (t, pose, k) => Stride(t, pose, 30f * k, 34f * k, 3f * k, 0.018f * k, ElbowRest)),
        ["run"] = new("run", 0.56f, true, false,
            (t, pose, k) => Stride(t, pose, 48f * k, 62f * k, 14f * k, 0.032f * k, -58f)),
    };

    /// <summary>Every clip name a game may play, procedural and keyed together.</summary>
    public static IReadOnlyList<string> Names { get; } =
        _procedural.Keys.Concat(_keyed.Keys).ToArray();

    /// <summary>Looks a clip up by name. Null for one that does not exist.</summary>
    public static ChibiClip? Find(string name)
        => _procedural.TryGetValue(name, out var procedural) ? procedural
            : _keyed.GetValueOrDefault(name);

    // -------------------------------------------------------------------------
    // Procedural locomotion
    // -------------------------------------------------------------------------

    private static void Idle(float t, ChibiPose pose, float intensity)
    {
        float b = MathF.Sin(t * Tau) * intensity;

        pose.Set("Torso", 1.5f + b * 1.6f, 0, 0);
        pose.Set("Head", -b * 2f, MathF.Sin(t * Tau * 0.5f) * 4f, 0);
        pose.Set("ArmL", b * 1.5f, 0, ArmRest + b);
        pose.Set("ArmR", b * 1.5f, 0, -ArmRest - b);
        pose.Set("ForearmL", ElbowRest, 0, 0);
        pose.Set("ForearmR", ElbowRest, 0, 0);
        pose.Offset = new Vector3(0, b * 0.006f, 0);
    }

    private static void Stride(float t, ChibiPose pose,
                               float swing, float knee, float lean, float bob, float elbow)
    {
        float s = MathF.Sin(t * Tau);
        float c = MathF.Cos(t * Tau);

        float thighL = s * swing;
        float thighR = -s * swing;
        // The knee only folds on the back half of the stride; a leg that bends going
        // forward is the single thing that makes a walk cycle look wrong.
        float shinL = -MathF.Max(0, -s) * knee;
        float shinR = -MathF.Max(0, s) * knee;

        pose.Set("ThighL", thighL, 0, 0);
        pose.Set("ThighR", thighR, 0, 0);
        pose.Set("ShinL", shinL, 0, 0);
        pose.Set("ShinR", shinR, 0, 0);
        // Feet stay roughly flat by undoing half of what the leg above them did.
        pose.Set("FootL", -(thighL + shinL) * 0.5f, 0, 0);
        pose.Set("FootR", -(thighR + shinR) * 0.5f, 0, 0);

        pose.Set("ArmL", -s * swing * 0.85f, 0, ArmRest);
        pose.Set("ArmR", s * swing * 0.85f, 0, -ArmRest);
        pose.Set("ForearmL", elbow, 0, 0);
        pose.Set("ForearmR", elbow, 0, 0);

        pose.Set("Hips", 0, -s * 6f, 0);
        pose.Set("Torso", lean, s * 5f, 0);
        pose.Set("Head", -lean * 0.5f, -s * 3f, 0);

        // Two bobs per stride: one for each foot that lands.
        pose.Offset = new Vector3(0, MathF.Abs(c) * bob, 0);
    }

    // -------------------------------------------------------------------------
    // Keyed clips
    // -------------------------------------------------------------------------

    private readonly record struct Key(float Time, Vector3 Value);

    private static Vector3 SampleTrack(Key[] keys, float t)
    {
        if (keys.Length == 0) return Vector3.Zero;
        if (t <= keys[0].Time) return keys[0].Value;

        for (int i = 1; i < keys.Length; i++)
        {
            if (t > keys[i].Time) continue;
            Key a = keys[i - 1];
            Key b = keys[i];
            float span = b.Time - a.Time;
            return Vector3.Lerp(a.Value, b.Value, span <= 0 ? 0 : (t - a.Time) / span);
        }
        return keys[^1].Value;
    }

    private static Dictionary<string, ChibiClip> LoadKeyed()
    {
        var clips = new Dictionary<string, ChibiClip>(StringComparer.Ordinal);
        try
        {
            using Stream? stream = typeof(ChibiClips).Assembly
                .GetManifestResourceStream("SexyBiscuit.Engine.chibi-clips.json");
            if (stream == null) return clips;

            using var document = JsonDocument.Parse(stream);
            if (!document.RootElement.TryGetProperty("clips", out JsonElement all)) return clips;

            foreach (JsonProperty entry in all.EnumerateObject())
            {
                JsonElement clip = entry.Value;
                var tracks = new Dictionary<string, Key[]>(StringComparer.Ordinal);

                if (clip.TryGetProperty("tracks", out JsonElement trackList))
                {
                    foreach (JsonProperty track in trackList.EnumerateObject())
                        tracks[track.Name] = ReadKeys(track.Value, "rot");
                }

                Key[] offset = clip.TryGetProperty("offset", out JsonElement o)
                    ? ReadKeys(o, "pos") : Array.Empty<Key>();

                clips[entry.Name] = new ChibiClip(
                    entry.Name,
                    clip.TryGetProperty("duration", out var d) ? (float)d.GetDouble() : 1f,
                    clip.TryGetProperty("loop", out var l) && l.GetBoolean(),
                    clip.TryGetProperty("hold", out var h) && h.GetBoolean(),
                    (t, pose, _) =>
                    {
                        foreach (var (joint, keys) in tracks) pose.Joints[joint] = SampleTrack(keys, t);
                        if (offset.Length > 0) pose.Offset = SampleTrack(offset, t);
                    });
            }
        }
        catch (Exception ex)
        {
            // Missing clips must not take the game down: locomotion still works, and the
            // reason is on the console rather than nowhere.
            Console.Error.WriteLine($"[ChibiClips] could not load the shared clips: {ex.Message}");
        }
        return clips;
    }

    private static Key[] ReadKeys(JsonElement array, string field)
    {
        var keys = new List<Key>();
        foreach (JsonElement key in array.EnumerateArray())
        {
            if (!key.TryGetProperty(field, out JsonElement value)) continue;
            keys.Add(new Key(
                (float)key.GetProperty("t").GetDouble(),
                new Vector3((float)value[0].GetDouble(), (float)value[1].GetDouble(),
                            (float)value[2].GetDouble())));
        }
        return keys.ToArray();
    }
}
